### PR TITLE
Best practices for using feature flag systems article

### PR DESCRIPTION
**Pull request by ferrantim**
## About the changes
Adding best practices for using feature flag systems article written with @thomasheartman, @chriswk  and @ivarconr.

### CODE REVIEW

**Comment by thomasheartman** [comment]
Nice! I'll have a proper look at this tomorrow, but a few things straight away:
1. We should put images along the rest of the doc images. That way, the docs don't need to reach out to github to fetch an image and redisplay it. Images go in `website/static/img/`
2. All images should have a useful alt text explaining what's in them. If the alt text is too long, which will typically be the case with diagrams, consider using a caption instead.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:59]:
```suggestion
**Example**: Do all members of Team A need updates on Project X, even if only one member of Team A is working on it? If so, include all team members in Project X’s group. Otherwise, only include the member who works on project X. 
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:63]:
```suggestion
While you should try to organize your flags based on high-level abstractions such as applications, teams, or business units, many services cut across these artificial organizing principles.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:91]:
```suggestion

The flag lifecycle stage dictates changes in the owners.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:242]:
```suggestion
Both feature flags and configuration settings control an application’s behavior. However, it's crucial to distinguish between feature flagging systems and configuration management systems and use each for its intended purpose.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:244]:
```suggestion
Feature flags are intended for dynamic and temporary use, giving you [runtime control](https://docs.getunleash.io/topics/feature-flags/runtime-control) to enable or disable functionality in a live environment. This flexibility supports [short-lived experiments](https://docs.getunleash.io/topics/feature-flags/short-lived-feature-flags), such as A/B testing or gradual rollouts, instead of relying on configuration settings that require an application redeploy to take effect.
```

I know this is a debate, but I saw examples where feature flags were great for things like configuring logs or putting an application into maintenance mode, which are in a way "a configuration."

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:248]:
```suggestion
A good rule of thumb is that if the data needs encryption, or contains PII (Personal Identifiable Information), it’s probably better suited for your Configuration management system than a feature flag system.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:326]:
```suggestion
#### Dependency on 3rd party services
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:330]:
```suggestion
#### Complexity and Maintainability
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:334]:
```suggestion
#### Performance Implications
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (alvinometric) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:338]:
```suggestion
#### Security Risks
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [reply website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:244]: This is a good point, that maybe we should change. I'd consider those "dynamic configuration", for which flags *are* great and not necessarily something we discourage. You might want to increase the amount of logging in a running app if something appears to be going wrong, and similarly you might want to turn on maintenance mode without restarting. I think those are entirely valid use cases and would say that flags are a good solution for that. 

So maybe we should just add those as good examples of using flags for *dynamic* configuration.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:155]: Assuming the images has moved to `/img/evaluate-flags-at-the-edge.png`

```suggestion
<Figure img="/img/evaluate-flags-at-the-edge.png" caption="Instead of evaluating a flag in a back-end service and then dynamically changing the behavior in that service (either by calling out to other services or changing how the service works), consider evaluating the flag as far out in the stack as you can and deciding which service to call based on the flag value."/>
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:186]: I think this image came out pretty nice in the end 😄 

Same thing as for previous images: move it to the docs repo and give it a caption. 

I'm assuming you've called it `/img/evaluate-flags-only-once.png`:

```suggestion
<Figure img="/img/evaluate-flags-only-once.png" caption="You should evaluate flags only once in a chain of services and pass the result on to the next service in the chain if it needs the value. If you evaluate the flag in each service, you risk the flag value changing between each service, causing unexpected results."/>
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [reply website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:251]: Agree with this. this feels short enough that a md table could work just as well. And if we wanna stick with html, we should consider reformatting this to make it easier to read with indentation etc.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:298]: This isn't a good link title because it requires context to understand what it points to. Consider something like this instead. Also, whenever you link to other docs, you should link directly to the file, not to the URL, as explained in the [docusaurus markdown links](https://docusaurus.io/docs/markdown-features/links) article. In particular, this means the build will break if the doc doesn't exist anymore instead of the build succeeding but the link giving you a 404.

```suggestion
We've said it before, but we'll say it again: [Feature flag names need to be globally unique](feature-flags-best-practices.md#8-use-unique-names-across-all-applications-enforce-naming-conventions). Every feature flag within the same Feature Flag Control service must have a unique name across the cluster to prevent inconsistencies and errors.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:300]: This intra-doc link should also be updated to point to the "physical" md file.
```suggestion
In a perfect world, all references to feature flags in code would be cleaned up as soon as the feature flag is archived.  But we don’t live in a perfect world. You will save yourself many hours of frustration and potential downtime if you use unique names to protect new features from being accidentally linked to outdated flags, which could inadvertently expose old features. [Unique names should ideally be enforced during flag creation](../../reference/feature-flag-naming-patterns.mdx).
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:350]: Update the link:

```suggestion
This guide was all about using feature flags at scale. For similar tips on building and scaling the feature flag system itself, see our [feature flag best practices guide](feature-flag-best-practices.md).
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:9]: The intra-doc link should point to the file.

```suggestion
This guide to using feature flag systems at scale is based on lessons learned working with some of the largest feature flag deployments in the world through the [Unleash Open-Source](https://github.com/Unleash/unleash) and [Enterprise Feature Management platform](https://www.getunleash.io/enterprise-feature-management-platform). However, the principles outlined do not apply only to Unleash. They apply to any large-scale feature flag system, including one you build yourself or another commercial solution.  For tips on how to stand up and run a feature flag system, see our [principles for building and scaling feature flags](feature-flag-best-practices.md).
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:67]:
```suggestion
This is why feature flag systems should be [open by default](feature-flag-best-practices.md#9-choose-open-by-default-democratize-feature-flag-access).  There are valid use cases for excluding flags from global search. For example, a public company in the middle of an acquisition where exposing flags related to the website or app changes might breach regulatory guidelines.  Your feature flag system should accommodate these private projects; however, they should be the exception, not the rule.
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:191]:
```suggestion
* Laws of distributed computing: Particularly in distributed systems, we cannot assume all parts of the system are perfectly synchronized, as networks are unreliable and can experience transient errors at any time. Feature flag systems generally [prefer availability over consistency](feature-flag-best-practices.md#6-design-for-failure-favor-availability-over-consistency). By only evaluating a feature flag once, we guarantee a consistent experience for our users. 
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:244]: the links here also need to be updated (as a separate point to Alvin's):

```suggestion
Feature flags are intended for dynamic and temporary use, giving you [runtime control](feature-flag-best-practices#1-enable-run-time-control-control-flags-dynamically-not-using-config-files) to enable or disable functionality in a live environment. This flexibility supports [short-lived experiments](feature-flag-best-practices#7-make-feature-flags-short-lived-do-not-confuse-flags-with-application-configuration), such as A/B testing or gradual rollouts, instead of relying on configuration settings that require an application redeploy to take effect. Despite the ease of turning off and on features or functionality with feature flags, they should not be repurposed for managing configuration.
```

**Review by thomasheartman — COMMENTED** [review]
Looks like Alvin's got you covered with formatting and content (🙌🏼), so I've had a look over focusing on image captions and links (wording and where they point to). Most places should have suggestions, but you'll still need to add the relevant images to the repo and double check that my link suggestions work.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (ferrantim) [reply website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md]: Gonna leave for now.  That is the spirit of the section and I think it is explained well, at least for me.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (ferrantim) [reply website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:252]: leaving as html for now.

**Comment by ferrantim** [comment]
@thomasheartman moved the images to the proper repo and accepted your suggestions. Can you take a look and approve if this looks good? Then I can merge. Thanks!

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:295]: The formatting here is off and it's breaking the build. Replacing with a new version.

```suggestion
<table>
  <tr>
    <th>Configuration system</th>
    <th>Feature flag system</th>
  </tr>
  <tr>
    <td>Is: Long-lived, static during runtime</td>
    <td>Is: Short-lived, changes during runtime</td>
  </tr>
  <tr>
    <td>
      Examples of things to update with the config system:
      <ul>
        <li>Database or server credentials</li>
        <li>Server port</li>
        <li>CORS headers</li>
        <li>API base URL</li>
        <li>SSO setup</li>
      </ul>
    </td>
    <td>
      Examples of things to update with feature flag system:
      <ul>
        <li>Dark mode / light mode</li>
        <li>Stable/Beta feature</li>
        <li>New or old algorithm</li>
        <li>A/B test</li>
      </ul>
    </td>
  </tr>
</table>
```

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:334]: I'd avoid using terms related to sanity in writing (for ableism reasons). Can we rephrase this to something else? What about this?

```suggestion
## Keeping cool while managing feature flags at scale
```

This also applies in the text, by the way.

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:43]:
```suggestion
<Figure img="/img/feature-flag-orginization-methods.png" caption="In situtations where predefined teams own a feature from start to finish, it makes sense to create a project for each team. In situtations where you have dynamic teams that form around feature work, you might instead want to create smaller projects for each feature, only giving access to the people who work on that feature."/>
```

Comment (thomasheartman) [reply website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:43]: @ferrantim You've put a typo in the file name of this one, so the image wasn't loading correctly. I can't easily change the file name, but I can update the reference. Not super important, but I suggest you go and update the file name and the reference later.

**Comment by thomasheartman** [comment]
I've had a look over again now. I've fixed the table, some images and added one more suggestion to wording (avoid using "sane" / "sanity"). 

I'm happy for you to merge it when you've addressed the comments, so I'll give you an approval.

But one thing: this doc isn't listed in the sidebar anywhere  (but you can preview it at https://unleash-docs-git-fork-ferrantim-patch-2-unleash-team.vercel.app/topics/feature-flags/best-practices-using-feature-flags-at-scale). This is because there's no entry for it in website/sidebars.js. You'll need to add it there for it to show up. Feel free to do that now or in a follow-up.

**Review by thomasheartman — APPROVED** [review]

**Comment by thomasheartman** [comment]
Oh, and maybe if it's feasible, we could use svgs instead of pngs for the images. Can you export as svg from excalidraw?

**Comment by ferrantim** [comment]
> Oh, and maybe if it's feasible, we could use svgs instead of pngs for the images. Can you export as svg from excalidraw?

I was able to swap 2 of the images that I had on my exaclidraw for svgs.  The one one on organizing feature flags that you created @thomasheartman I didn't have. Can you created a new PR to update that once this merged?

**Comment by ferrantim** [comment]
Actually @thomasheartman looks like I don't have permission to merge. Can you do that please?

File: website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md
Comment (thomasheartman) [inline website/docs/topics/feature-flags/best-practices-using-feature-flags-at-scale.md:336]:
```suggestion
These best practices come from working with many of the largest organizations on the planet. We’ve learned that keeping cool while managing feature flags at scale requires careful planning, thoughtful organization, and adherence to several best practices. By aligning feature flags with organizational structure, making flags searchable, and managing permissions effectively, you can maintain a clean and efficient system that sets you up for success from the start. Instrumenting your code correctly and regularly paying down technical debt helps maintain performance and security. Finally, avoiding common pitfalls, such as using feature flags for business logic or configuration management, keeps your system robust and manageable.
```

**Comment by alvinometric** [comment]
@thomasheartman Can you add @ferrantim to the repo as a member?

**Comment by thomasheartman** [comment]
I'm traveling right now, but I'll see what I can do later

**Pull request merged**
